pkg/driver/aws-ebs/aws_ebs_tags_controller.go

```diff
@@ -0,0 +1,441 @@
+func (c *EBSVolumeTagsController) processInfrastructure(ctx context.Context, infra *configv1.Infrastructure, ec2Client *ec2.EC2) error {
+	if infra.Status.PlatformStatus != nil && infra.Status.PlatformStatus.AWS != nil &&
+		infra.Status.PlatformStatus.AWS.ResourceTags != nil {
+		awsInfra := infra.Status.PlatformStatus.AWS
+		err := c.fetchPVsAndUpdateTags(ctx, awsInfra.ResourceTags, ec2Client)
+		if err != nil {
+			klog.Errorf("Error processing PVs for infrastructure update: %v", err)
+			return err
+		}
+	}
+	return nil
+}
+
+// fetchPVsAndUpdateTags retrieves all PVs and updates the AWS EBS tags in batches of 100
+func (c *EBSVolumeTagsController) fetchPVsAndUpdateTags(ctx context.Context, resourceTags []configv1.AWSResourceTag, ec2Client *ec2.EC2) error {
+	pvs, err := c.listPersistentVolumesWithRetry(ctx)
+	if err != nil {
+		return fmt.Errorf("error fetching PVs: %v", err)
+	}
+	// Compute the hash for the new set of tags
+	newTagsHash := computeTagsHash(resourceTags)
+	pvsToBeUpdated := filterUpdatableVolumes(pvs, newTagsHash)
+
+	// If there are no volumes to update, return early
+	if len(pvsToBeUpdated) == 0 {
+		klog.Infof("No volume tags to update as hashes are unchanged")
+		return nil
+	}
+
+	// Process the volumes in batches
+	for i := 0; i < len(pvsToBeUpdated); i += batchSize {
+		end := i + batchSize
+		if end > len(pvsToBeUpdated) {
+			end = len(pvsToBeUpdated)
+		}
+		batch := pvsToBeUpdated[i:end]
+
+		// Update tags on AWS EBS volumes
+		err = c.updateEBSTags(batch, ec2Client, resourceTags)
+		if err != nil {
+			c.handleTagUpdateFailure(batch, err)
+			continue
+		}
+
+		// Update PV annotations after successfully updating the tags in AWS
+		for _, volume := range batch {
+			// Set the new tag hash annotation in the PV object
+			setPVTagHash(volume, newTagsHash)
+
+			// Update the PV with the new annotations
+			err = c.updateVolumeWithRetry(ctx, volume)
+			if err != nil {
+				klog.Errorf("Error updating PV annotations for volume %s: %v", volume.Name, err)
+				c.failedQueue.AddRateLimited(volume.Name) // Retry updating annotation if update fails
+				continue
+			}
+			klog.Infof("Successfully updated PV annotations and tags for volume %s", volume.Name)
+		}
+	}
+	return nil
+}
+
+// updateEBSTags updates the tags of an AWS EBS volume
+func (c *EBSVolumeTagsController) updateEBSTags(pvBatch []*v1.PersistentVolume, ec2Client *ec2.EC2, resourceTags []configv1.AWSResourceTag) error {
+
+	// Merge the existing tags with new resource tags
+	mergedTags := newAndUpdatedTags(resourceTags)
+
+	// Create or update the tags
+	_, err := ec2Client.CreateTags(&ec2.CreateTagsInput{
+		Resources: pvsToResourceIDs(pvBatch),
+		Tags:      mergedTags,
+	})
+
+	if err != nil {
+		return fmt.Errorf("error creating tags for volume %v: %v", pvBatch, err)
+	}
+	return nil
+}
+
+func (c *EBSVolumeTagsController) listPersistentVolumesWithRetry(ctx context.Context) ([]*v1.PersistentVolume, error) {
```

Comment:
same comment as before. There is no need to list PV objects from a informer with an exponential backoff. Informer has cached objects, it isn't supposed to fail.